cloudfoundry-incubator/kubo-release
# Enable the metrics-server

**What this PR does / why we need it**:
This PR enables the `metrics-server` and optionally allows the operator to delete 
the existing `Heapster` deployment. `Heapster` is being deprecated as of K8s 1.11
and has not been maintained with the `metrics.k8s.io/v1beta1` APIs.

By enabling the `metrics-server`, we're also enabling the [Horizontal Pod Autoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/), which consumes the `metrics.k8s.io/v1beta1` APIs. In addition, `kubectl top` will now work out of the box without needing to set `--heapster-scheme=https`.

This PR also cleans up the `apply-addons/specs` directory a little bit, moving each addon
to its own directory so that we can globally apply all specs in the directory. Heapster and
Influxdb, though no longer supported, are kept around for two reasons:

* Kubernetes Dashboard does not yet support the metrics API
* Cleanup of Heapster and Influxdb (`kubectl delete -f <file>`)

Heapster and Influxdb (and their related configurations) can be cleaned up with `delete-heapster` property on the `apply-addons` job.

`metrics-server` specs are based off the incubator branch [seen here](https://github.com/kubernetes-incubator/metrics-server/tree/master/deploy/1.8+).
I had to add an extra RBAC rule due to our secured Kubelet and then configured the deployment with certs for secure serving.

**How can this PR be verified?**
There are spec tests in this PR that verify the rendered templates in the `apply_addons` job.
[Horizontal Pod Autoscaler Walkthrough](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/) provides a step-by-step example of the HPA in action.

**Is there any change in kubo-deployment?**
https://github.com/cloudfoundry-incubator/kubo-deployment/pull/315

**Is there any change in kubo-ci?**
Not at the moment. There should be tests that ensure the HPA work, however.

**Does this affect upgrade, or is there any migration required?**
Not unless the operator specifies `delete-heapster: true`, then those services will be deleted and
Dashboard will not properly serve the pod metrics.

**Which issue(s) this PR fixes:**
[#157915809](https://www.pivotaltracker.com/story/show/157915809)

**Release note**:

```release-note
Enabled the `metrics-server` and registered it as an API for `metrics.k8s.io/v1beta1`. 
This, in turn, enables certain metrics-based features like the Horizontal Pod Autoscaler
and a better integration with features like `kubectl top`.
```
